src/codegen.cpp

```diff
@@ -494,6 +495,57 @@ static inline void add_named_global(JuliaVariable *name, void *addr)
     add_named_global(name->name, addr);
 }
 
+static AttributeList get_attrs_ipoeffects(LLVMContext &context, std::pair<bool,uint32_t> effect, bool is_decl)
+{
+    if (!effect.first)
+        return AttributeList();
+    uint32_t effects = effect.second;
+    uint8_t consistent = (effects >> 0) & 0x07;
+    uint8_t effect_free = (effects >> 3) & 0x03;
+    bool nothrow = (effects >> 5) & 0x01;
+    bool terminates = (effects >> 6) & 0x01;
+    bool notaskstate = (effects >> 7) & 0x01;
+    uint8_t inaccessiblememonly = (effects >> 8) & 0x03;
+    bool nonoverlayed = (effects >> 10) & 0x01;
+    bool noinbounds = (effects >> 11) & 0x01;
+    AttrBuilder attr(context);
+
+    if (consistent == 0) {
+        attr.addAttribute("consistent");
+    }
+    if (effect_free == 0) {
+        attr.addAttribute("effect_free");
+    }
+    if (nothrow == 1) {
+        attr.addAttribute("nothrow");
+        attr.addAttribute(Attribute::NoUnwind);
+    }
+    if (terminates == 1){
+        attr.addAttribute("terminates");
+        attr.addAttribute(Attribute::WillReturn);
+    }
+    if (inaccessiblememonly == 0) {
+        attr.addAttribute("inaccessiblememonly");
+        attr.addAttribute(Attribute::InaccessibleMemOnly);
+    } else if (inaccessiblememonly == 2){
+        attr.addAttribute("inaccessiblememorargmemonly");
+        attr.addAttribute(Attribute::InaccessibleMemOrArgMemOnly);
+    }
+    if (notaskstate == 1) {
+        attr.addAttribute("notaskstate");
+    }
+    if (nonoverlayed == 1) {
+        attr.addAttribute("nonoverlayed");
+    }
+    if (noinbounds == 1) {
+        attr.addAttribute("noinbounds");
+    }
+    if (is_decl && consistent == 0 && effect_free == 0 && nothrow == 1 && terminates == 1){
+        attr.addAttribute(Attribute::Speculatable);
```

Comment:
You probably want to change these to be added at functionindex

It may be worthwhile considering if any attributes on params are also useful

Reply:
I think they might be, but I'm not sure what information we can infer from here

Reply:
Also when creating the AttrList we pass the builder to the `FnAttr` builder